### PR TITLE
4.12 rebase

### PR DESCRIPTION
**Pull request by poechsel**
Merge changes from 4.12

4.12 was taken at revision 7e6b22b232146f9177cbfa874b3fcc66ceed192b

### CODE REVIEW

File: asmcomp/cmm.mli [hunk 1]
```diff
@@ -176,17 +176,12 @@ type memory_chunk =
...
-      { func : string; ty : machtype;
-        alloc : bool ; label_after : label option;
+      { func : string;
+        ty : machtype;
+        ty_args : exttype list;
+        alloc : bool;
         returns : bool; }
-=======
-  | Cextcall of string * machtype * exttype list * bool
-      (** The [machtype] is the machine type of the result.
```
Comment (mshinwell) [inline asmcomp/cmm.mli]: Looks like this comment has been lost

File: middle_end/closure/closure.ml [hunk 2]
```diff
@@ -798,15 +798,16 @@ let direct_apply env fundesc ufunct uargs ~loc ~attribute =
...
 
 let strengthen_approx appl approx =
   match approx_ulam appl with
     (Value_const _) as intapprox ->
       intapprox
   | _ -> approx
 
-(* If a term has approximation Value_integer and is pure,
+(* If a term has approximation Value_integer or Value_constptr and is pure,
    replace it by an integer constant *)
```
Comment (mshinwell) [inline middle_end/closure/closure.ml]: I think this is the wrong resolution (likewise for lines 801/802 above).  We should be removing the "or Value_constptr" phrase.

File: middle_end/closure/closure.ml [hunk 3]
```diff
@@ -1071,15 +1070,12 @@ let rec close ({ backend; fenv; cenv ; mutable_vars } as env) lam =
...
-      close env
-        (Lapply{
-           ap_loc=loc;
-           ap_func=funct;
-           ap_args=[arg];
-           ap_tailcall=Default_tailcall;
-           ap_inlined=Default_inline;
-           ap_specialised=Default_specialise;
-         })
+      close env       (Lapply{ap_loc=loc;
```
Comment (mshinwell) [inline middle_end/closure/closure.ml]: Bad formatting

Comment (mshinwell) [reply middle_end/closure/closure.ml]: (in fact I'm not sure there should be any change here at all)

File: middle_end/flambda/to_cmm/un_cps_helper.ml [hunk 4]
```diff
@@ -28,11 +28,23 @@ let arch64 = Arch.size_int = 8
...
+  | Value -> Cmm.XInt
+  | Naked_number Naked_float -> Cmm.XFloat
+  | Naked_number Naked_int64 -> Cmm.XInt64
+  | Naked_number Naked_int32 -> Cmm.XInt32
+  | Naked_number (Naked_immediate | Naked_nativeint) ->
+    begin match Targetint.num_bits with
+    | Thirty_two -> Cmm.XInt32
+    | Sixty_four -> Cmm.XInt64
+    end
+  | Fabricated -> Misc.fatal_error "[Fabricated] king not expected here"
```
Comment (mshinwell) [inline middle_end/flambda/to_cmm/un_cps_helper.ml]: king -> kind

File: middle_end/flambda/to_cmm/un_cps_helper.mli [hunk 5]
```diff
@@ -27,6 +27,7 @@ val arch64 : bool
 val typ_int64 : Cmm.machtype
 (** An adequate Cmm machtype for an int64 (including on a 32-bit target). *)
 
+val  exttype_of_kind : Flambda_kind.t -> Cmm.exttype
```
Comment (mshinwell) [inline middle_end/flambda/to_cmm/un_cps_helper.mli]: Double space after "val"

File: runtime/compact.c [hunk 6]
```diff
@@ -15,6 +15,7 @@
 
 #define CAML_INTERNALS
 
+
```
Comment (mshinwell) [inline runtime/compact.c]: Unnecessary change

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Comment by mshinwell** [comment]
A few minor things to fix.  Then please try building Coq with this PR.  (I'm not sure exactly which version of Coq is needed, but I believe there does exist one that supports OCaml 4.12.)

File: middle_end/closure/closure.ml (hunk 2 above)
Comment (poechsel) [reply middle_end/closure/closure.ml]: My bad, I forgot to edit the comments when removing Value_constptr

**Comment by poechsel** [comment]
Coq is building successfully with this PR.

**Comment by lthls** [comment]
Out of curiosity, why did you need to remove the comparison of bytecode programs in `ocamltest` ? I'd not expect that to be necessary.

**Comment by poechsel** [comment]
Some tests were failing on the comparison of bytecode programs. Upon closer inspection we noticed that the bytecode was identical at the exception of the digests for one cmi file. We then looked at the cmi files generated using ocamlc.opt and ocamlc.byte and noticed that their content were identical, but when was marshalled one of {ocamlc.opt, ocamlc.byte} was sharing more things than the other one, thus leading to the test failing. We were not able to find the root cause causing the discrepancies in sharing so we decided to deactivate them as was done in https://github.com/ocaml/ocaml/pull/9677

**Review by mshinwell — APPROVED** [review]

**Comment by Gbury** [comment]
This PR seems to have brought the "stale issue" bot from upstream, was that intended ?

**Comment by poechsel** [comment]
Not really

**Pull request merged**